lib/facility.rb

```diff
@@ -1,14 +1,188 @@
+    facility_data.map do |data|
+      name = Helper.title_case(data[:name])
+      street = Helper.title_case(data[:address1])
+      city = Helper.title_case(data[:city])
+      state = data[:state]
+      zip_code = data[:zipcode]
+      phone = data[:phone]
+      website = data[:facebook_url]
+      address = "#{street}, #{city}, #{state}, #{zip_code}" # Concatenate address components
+
+      Facility.new(name: name, address: address, phone: phone, website: website)
+    end
   end
 
-  def add_services(service)
+  # Adds a service to the facility
+  def add_service(service)
     @services << service
   end
+
+  # Registers a vehicle if the facility offers the service
+  # and the vehicle is not already registered
+  def register_vehicle(vehicle)
+    raise 'Vehicle already registered.' if @registered_vehicles.include?(vehicle)
+
+    return nil unless @services.include?('Vehicle Registration')
+
+    vehicle.registration_date = Date.today
+    vehicle.determine_plate_type
+    @registered_vehicles << vehicle
+    calculate_collected_fees(vehicle)
+    @registered_vehicles
+  end
+
+  # Calculates the fees collected for a vehicle registration
+  def calculate_collected_fees(vehicle)
+    plate_type_fees = {
+      ev: 200,
+      antique: 25
+    }
+
+    @collected_fees += plate_type_fees[vehicle.plate_type] || 100
+  end
+
+  # Registers a registrant if the facility offers the service and the registrant doesn't already
+  # have a permit.
+  def earn_permit(registrant)
+    raise 'Already has permit.' if registrant.permit?
+
+    return nil unless registrant.age > 15
+
+    registrant.earn_permit
+  end
+
+  # Administers a written test if the facility offers the service,
+  # if the registrant has a permit, if the registrant is over 15,
+  # and has not already taken a written test.
+  def administer_written_test(registrant)
+    raise 'Already took written test.' if registrant.written_test?
+
+    return nil unless registrant.permit? && registrant.age > 15 && @services.include?('Written Test')
+
+    registrant.administer_written_test
+  end
+
+  # Administers a road test if the facility offers the service,
+  # if the registrant took the written test,
+  # and if the registrant has not already taken a road test.
+  def administer_road_test(registrant)
+    raise 'Already took road test and has license.' if registrant.road_test?
+
+    return nil unless registrant.written_test? && @services.include?('Road Test')
+
+    registrant.administer_road_test
+  end
+
+  # Renews a drivers license if the facility offers the service,
+  # if the registrant has taken the written test and road test,
+  # and if the registrant has not already renewed their license.
+  def renew_drivers_license(registrant)
+    raise 'Already renewed drivers license.' if registrant.renewed_drivers_license?
+
+    return nil unless registrant.written_test? && registrant.road_test? && services.include?('Renew License')
+
+    registrant.renew_drivers_license
+  end
```

Comment:
Great job overall with breaking up methods into smaller helper methods. It really helps with readability!